File: web/src/lib/pricing.ts
// Rough Anthropic model pricing per 1M tokens. Public list pricing as
// of 2026; surfaced as "approx" in the UI so users know to trust the
// invoice for the authoritative number. Update the table as Anthropic
// publishes new model rates.

type Rate = { input: number; output: number };

// Provider rate tables. Patterns match family identifiers, not
// specific versions — providers re-use the same tier across versions
// inside a family. Update as providers publish new rates.
const ANTHROPIC_RATES: Array<{ pattern: RegExp; rate: Rate }> = [
  // Fable 5 (Mythos-class) is the premium tier — list it first; it doesn't
  // overlap the opus/sonnet/haiku patterns. $10/$50 per MTok (GA 2026-06-09).
  { pattern: /claude-fable/i, rate: { input: 10, output: 50 } },
  // Current Opus (4.5–4.8) is $5/$25. The only $15/$75 Opus models were 4.1
  // and 4.0, both deprecated and retiring mid-2026 — not worth special-casing.
  { pattern: /claude-opus/i, rate: { input: 5, output: 25 } },
  { pattern: /claude-sonnet/i, rate: { input: 3, output: 15 } },
  { pattern: /claude-haiku/i, rate: { input: 1, output: 5 } },
];

// OpenAI public list pricing per 1M tokens, by model family.
// Source: https://platform.openai.com/docs/pricing
const OPENAI_RATES: Array<{ pattern: RegExp; rate: Rate }> = [
  // GPT-5 family (current premium tier).
  { pattern: /^gpt-5-mini/i, rate: { input: 0.25, output: 2 } },
  { pattern: /^gpt-5-nano/i, rate: { input: 0.05, output: 0.4 } },
  { pattern: /^gpt-5/i, rate: { input: 1.25, output: 10 } },
  // GPT-4o family.
  { pattern: /^gpt-4o-mini/i, rate: { input: 0.15, output: 0.6 } },
  { pattern: /^gpt-4o/i, rate: { input: 2.5, output: 10 } },
  // GPT-4.1 family.
  { pattern: /^gpt-4\.1-nano/i, rate: { input: 0.1, output: 0.4 } },
  { pattern: /^gpt-4\.1-mini/i, rate: { input: 0.4, output: 1.6 } },
  { pattern: /^gpt-4\.1/i, rate: { input: 2, output: 8 } },
  // o-series reasoning.
  { pattern: /^o3-mini/i, rate: { input: 1.1, output: 4.4 } },
  { pattern: /^o3/i, rate: { input: 2, output: 8 } },
];

/**
 * Returns the estimated USD cost of a run, or null if we don't have a
 * pricing entry for the model. Model strings look like
 * `provider:model-name` (e.g. `anthropic:claude-sonnet-4-6`,
 * `openai:gpt-4o-mini`).
 */
function lookupRate(model: string): Rate | null {
  const tables: Array<{ prefix: string; rates: typeof ANTHROPIC_RATES }> = [
    { prefix: "anthropic:", rates: ANTHROPIC_RATES },
    { prefix: "openai:", rates: OPENAI_RATES },
  ];
  for (const { prefix, rates } of tables) {
    if (!model.startsWith(prefix)) continue;
    const modelName = model.slice(prefix.length);
    return rates.find((r) => r.pattern.test(modelName))?.rate ?? null;
  }
  return null;
}

export function estimateRunCost(
  model: string,
  tokensInput: number,
  tokensOutput: number,
): number | null {
  const rate = lookupRate(model);
  if (!rate) return null;
  return (tokensInput * rate.input + tokensOutput * rate.output) / 1_000_000;
}

/** Cost of just the input or just the output tokens, for per-direction display. */
export function estimateTokenCost(
  model: string,
  tokens: number,
  direction: "input" | "output",
): number | null {
  const rate = lookupRate(model);
  if (!rate) return null;
  return (tokens * rate[direction]) / 1_000_000;
}

export function formatTokens(n: number): string {
  return new Intl.NumberFormat("en-US").format(n);
}

/** 3-significant-figure abbreviation: 9_502 → "9.50k", 15_100 → "15.1k",
 *  152_000 → "152k", 1_520_000 → "1.52M". Sub-1000 shows the plain number. */
export function abbreviateTokens(n: number): string {
  if (n < 1000) return String(n);
  for (const [div, unit] of [
    [1_000_000_000, "B"],
    [1_000_000, "M"],
    [1_000, "k"],
  ] as const) {
    if (n >= div) {
      const v = n / div;
      const s = v >= 100 ? Math.round(v).toString() : v >= 10 ? v.toFixed(1) : v.toFixed(2);
      return `${s}${unit}`;
    }
  }
  return String(n);
}

/** Cost rounded to the nearest penny, leading zero dropped: 0.048 → "$.05". */
export function formatPenny(usd: number): string {
  return `$${usd.toFixed(2).replace(/^0(?=\.)/, "")}`;
}

export function formatCurrency(usd: number): string {
  // Sub-cent values aren't useful at this UI level; floor to two
  // decimals but bump to three when the cost is below $0.01 so users
  // see something other than "$0.00".
  const decimals = usd < 0.01 ? 4 : usd < 1 ? 3 : 2;
  return new Intl.NumberFormat("en-US", {
    style: "currency",
    currency: "USD",
    minimumFractionDigits: decimals,
    maximumFractionDigits: decimals,
  }).format(usd);
}

File: api/src/pricing.rs
//! Per-million-token list pricing for the Anthropic + OpenAI model
//! families we support. Mirrors web/src/lib/pricing.ts — keep the
//! tables in lockstep when adding a new provider/family.
//!
//! Estimates only — the authoritative number is whatever each
//! provider invoices. The UI surfaces this as "approx" so the user
//! knows to trust their bill for the real number.

use once_cell::sync::Lazy;
use regex::Regex;

struct ModelRate {
    pattern: Regex,
    input_per_million: f64,
    output_per_million: f64,
}

static ANTHROPIC_RATES: Lazy<Vec<ModelRate>> = Lazy::new(|| {
    vec![
        // Fable 5 (Mythos-class) — premium tier, $10/$50 per MTok (GA 2026-06-09).
        rate("claude-fable", 10.0, 50.0),
        // Current Opus (4.5–4.8) is $5/$25; only the deprecated 4.1/4.0 were
        // $15/$75 and they retire mid-2026 — not worth special-casing.
        rate("claude-opus", 5.0, 25.0),
        rate("claude-sonnet", 3.0, 15.0),
        rate("claude-haiku", 1.0, 5.0),
    ]
});

static OPENAI_RATES: Lazy<Vec<ModelRate>> = Lazy::new(|| {
    vec![
        // GPT-5.x — specific patterns before the bare `^gpt-5` catch-all so it
        // doesn't swallow the 5.x variants. 5.5/5.4 are the current flagships.
        rate("^gpt-5\\.5", 5.0, 30.0),
        rate("^gpt-5\\.4-mini", 0.75, 4.5),
        rate("^gpt-5\\.4-nano", 0.2, 1.25),
        rate("^gpt-5\\.4", 2.5, 15.0),
        rate("^gpt-5\\.2", 0.875, 7.0),
        rate("^gpt-5\\.1", 0.625, 5.0),
        rate("^gpt-5-mini", 0.25, 2.0),
        rate("^gpt-5-nano", 0.05, 0.4),
        rate("^gpt-5", 1.25, 10.0),
        rate("^gpt-4o-mini", 0.15, 0.6),
        rate("^gpt-4o", 2.5, 10.0),
        rate("^gpt-4\\.1-nano", 0.1, 0.4),
        rate("^gpt-4\\.1-mini", 0.4, 1.6),
        rate("^gpt-4\\.1", 2.0, 8.0),
        rate("^o3-mini", 1.1, 4.4),
        rate("^o3", 2.0, 8.0),
    ]
});

fn rate(pattern: &str, input: f64, output: f64) -> ModelRate {
    ModelRate {
        // The web-side table is case-insensitive (`/…/i`); mirror that.
        pattern: Regex::new(&format!("(?i){pattern}"))
            .expect("invalid hard-coded model pricing regex"),
        input_per_million: input,
        output_per_million: output,
    }
}

// Prompt-cache multipliers on the base INPUT rate. Anthropic bills cache
// writes (creation) at 1.25x and cache reads at 0.1x of input. `tokens_input`
// is the uncached input (full rate); cache halves are counted separately and
// never overlap it. OpenAI/others that don't report cache tokens pass 0 here,
// so the multipliers are a no-op for them.
const CACHE_READ_MULTIPLIER: f64 = 0.1;
const CACHE_WRITE_MULTIPLIER: f64 = 1.25;

/// Returns USD cost for a run, or None if the model isn't in our
/// pricing tables (callers persist None which renders as "—" in
/// the UI). Model strings look like `provider:model-name` —
/// matching the same shape the TS-side helper consumes.
/// `cache_read`/`cache_write` are the Anthropic prompt-cache token counts
/// (0 when caching is off).
pub fn estimate_run_cost(
    model: &str,
    tokens_input: i32,
    tokens_output: i32,
    cache_read: i32,
    cache_write: i32,
) -> Option<f64> {
    if let Some(rest) = model.strip_prefix("anthropic:") {
        cost_in_table(rest, tokens_input, tokens_output, cache_read, cache_write, &ANTHROPIC_RATES)
    } else if let Some(rest) = model.strip_prefix("openai:") {
        cost_in_table(rest, tokens_input, tokens_output, cache_read, cache_write, &OPENAI_RATES)
    } else {
        None
    }
}

fn cost_in_table(
    model_name: &str,
    tokens_input: i32,
    tokens_output: i32,
    cache_read: i32,
    cache_write: i32,
    table: &Lazy<Vec<ModelRate>>,
) -> Option<f64> {
    let entry = table.iter().find(|r| r.pattern.is_match(model_name))?;
    let input_cost = (tokens_input as f64
        + cache_read as f64 * CACHE_READ_MULTIPLIER
        + cache_write as f64 * CACHE_WRITE_MULTIPLIER)
        * entry.input_per_million;
    let cost = (input_cost + tokens_output as f64 * entry.output_per_million) / 1_000_000.0;
    Some(cost)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn anthropic_sonnet() {
        let c =
            estimate_run_cost("anthropic:claude-sonnet-4-6", 1_000_000, 1_000_000, 0, 0).expect("priced");
        // 3 + 15 = 18 USD for 1M in + 1M out
        assert!((c - 18.0).abs() < 1e-9, "expected ~18.0, got {c}");
    }

    #[test]
    fn anthropic_opus() {
        let c =
            estimate_run_cost("anthropic:claude-opus-4-8", 1_000_000, 1_000_000, 0, 0).expect("priced");
        // 5 + 25 = 30 USD for 1M in + 1M out
        assert!((c - 30.0).abs() < 1e-9, "expected ~30.0, got {c}");
    }

    #[test]
    fn anthropic_fable_5() {
        let c =
            estimate_run_cost("anthropic:claude-fable-5", 1_000_000, 1_000_000, 0, 0).expect("priced");
        // 10 + 50 = 60 USD for 1M in + 1M out
        assert!((c - 60.0).abs() < 1e-9, "expected ~60.0, got {c}");
    }

    #[test]
    fn openai_gpt_4o_mini() {
        let c = estimate_run_cost("openai:gpt-4o-mini", 1_000_000, 1_000_000, 0, 0).expect("priced");
        // 0.15 + 0.6 = 0.75
        assert!((c - 0.75).abs() < 1e-9, "expected ~0.75, got {c}");
    }

    #[test]
    fn openai_gpt_5_5_not_swallowed_by_catch_all() {
        // gpt-5.5 must hit its own $5/$30 rate, not the bare ^gpt-5 ($1.25/$10).
        let c = estimate_run_cost("openai:gpt-5.5", 1_000_000, 1_000_000, 0, 0).expect("priced");
        // 5 + 30 = 35
        assert!((c - 35.0).abs() < 1e-9, "expected ~35.0, got {c}");
    }

    #[test]
    fn openai_gpt_5_base() {
        let c = estimate_run_cost("openai:gpt-5", 1_000_000, 1_000_000, 0, 0).expect("priced");
        // 1.25 + 10 = 11.25
        assert!((c - 11.25).abs() < 1e-9, "expected ~11.25, got {c}");
    }

    #[test]
    fn anthropic_cache_tokens_priced_at_read_write_multipliers() {
        // Opus input $5/MTok. 1M cache_read @ 0.1x = $0.50; 1M cache_write @
        // 1.25x = $6.25; 1M uncached input @ 1x = $5; no output.
        let c = estimate_run_cost(
            "anthropic:claude-opus-4-8",
            1_000_000,
            0,
            1_000_000,
            1_000_000,
        )
        .expect("priced");
        assert!((c - (5.0 + 0.5 + 6.25)).abs() < 1e-9, "expected ~11.75, got {c}");
    }

    #[test]
    fn unknown_model_returns_none() {
        assert!(estimate_run_cost("custom:no-such-model", 100, 100, 0, 0).is_none());
    }

    #[test]
    fn unknown_provider_returns_none() {
        assert!(estimate_run_cost("cohere:command", 100, 100, 0, 0).is_none());
    }
}
